python/repeated.c

```diff
@@ -0,0 +1,705 @@
+    // This is a set, not a delete.
+    upb_arena* arena = PyUpb_Arena_Get(self->arena);
+    if (upb_fielddef_issubmsg(f)) {
+      PyErr_SetString(PyExc_TypeError, "does not support assignment");
+      return -1;
+    }
+    if (step == 0) {
+      upb_msgval msgval;
+      if (!PyUpb_PyToUpb(value, f, &msgval, arena)) return -1;
+      upb_array_set(arr, idx, msgval);
+      return 0;
+    } else {
+      PyObject* seq =
+          PySequence_Fast(value, "must assign iterable to extended slice");
+      if (!seq) return -1;
+      if (PySequence_Size(seq) != count) {
+        PyErr_Format(PyExc_ValueError,
+                     "attempt to assign sequence of size %zd to extended slice "
+                     "of size %zd",
+                     PySequence_Size(seq), count);
+        Py_DECREF(seq);
+        return -1;
+      }
+      for (Py_ssize_t i = 0; i < count; i++, idx += step) {
+        upb_msgval msgval;
+        PyObject* item = PySequence_GetItem(seq, i);
+        // XXX: if this fails we can leave the list partially mutated.
+        bool ok = PyUpb_PyToUpb(item, f, &msgval, arena);
+        Py_DECREF(item);
+        if (!ok) return -1;
+        upb_array_set(arr, idx, msgval);
+      }
+    }
+  } else {
+    // This is a delete, not a set.
+    // Normalize direction: deletion is order-independent.
+    if (step > 0) {
+      idx += step * (count - 1);
+      step = -step;
+    }
+    if (step == -1) {
+      // Contiguous range.
+      upb_array_delete(arr, idx - (count - 1), count);
+    } else {
+      // Stepped slice.
+      for (Py_ssize_t i = 0; i < count; i++, idx += step) {
+        upb_array_delete(arr, idx, 1);
+      }
+    }
+  }
+  return 0;
+}
+
+static PyObject* PyUpb_RepeatedContainer_Pop(PyObject* _self, PyObject* args) {
+  PyUpb_RepeatedContainer* self = (PyUpb_RepeatedContainer*)_self;
+  Py_ssize_t index = -1;
+  if (!PyArg_ParseTuple(args, "|n", &index)) return NULL;
+  upb_array* arr = PyUpb_RepeatedContainer_AssureWritable(self);
+  size_t size = upb_array_size(arr);
+  if (index < 0) index += size;
+  if (index >= size) index = size - 1;
+  PyObject* ret = PyUpb_RepeatedContainer_Item(_self, index);
+  if (!ret) return NULL;
+  upb_array_delete(self->ptr.arr, index, 1);
+  return ret;
+}
+
+static bool PyUpb_RepeatedContainer_Assign(PyObject* _self, PyObject* list) {
+  PyUpb_RepeatedContainer* self = (PyUpb_RepeatedContainer*)_self;
+  const upb_fielddef* f = PyUpb_RepeatedContainer_GetField(self);
+  upb_array* arr = PyUpb_RepeatedContainer_AssureWritable(self);
+  Py_ssize_t size = PyList_Size(list);
+  bool submsg = upb_fielddef_issubmsg(f);
+  upb_arena* arena = PyUpb_Arena_Get(self->arena);
+  for (Py_ssize_t i = 0; i < size; ++i) {
+    PyObject* obj = PyList_GetItem(list, i);
+    upb_msgval msgval;
+    if (submsg) {
+      msgval.msg_val = PyUpb_CMessage_GetIfWritable(obj);
+      assert(msgval.msg_val);
+    } else {
+      if (!PyUpb_PyToUpb(obj, f, &msgval, arena)) return false;
```

Comment:
it feels weird that this can fail halfway through assigning the items in a list...